### PR TITLE
Flow-Regime MoE Output Routing: expert specialization by config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current model uses a single output projection for all flow regimes (single-foil, tandem, high-Re). Different regimes have qualitatively different pressure distributions — tandem interference creates wakes and downwash absent in single-foil, while high-Re adds turbulent boundary layer effects. A lightweight Mixture-of-Experts at the output stage lets the model specialize.

**Key distinction from failed MoE FFN routing:** That approach used hard dispatch of tokens to FFN experts (starved experts in small-dataset regime). This approach uses SOFT blending of output-stage expert heads, conditioned on global flow regime features. The full Transolver backbone is shared — only the final output projection is split.

**Reference:** "Mixture of Experts for External Aerodynamics" (arXiv:2508.21249, 2025). Reports 8-15% improvement on regime boundaries.

## Instructions

Replace the final output linear layer with a 3-expert soft-routed MoE head.

### Step 1: Add arguments

```python
parser.add_argument('--moe_output', action='store_true')
parser.add_argument('--moe_n_experts', type=int, default=3)
parser.add_argument('--moe_balance_weight', type=float, default=0.01,
                    help='Load-balancing aux loss weight')
```

### Step 2: Implement RegimeMoE module

```python
class RegimeMoE(nn.Module):
    def __init__(self, hidden_dim, n_experts=3, out_dim=3):
        super().__init__()
        self.gate = nn.Linear(4, n_experts)  # input: [re_norm, aoa_norm, is_tandem, gap_norm]
        nn.init.zeros_(self.gate.weight)  # uniform initial routing
        nn.init.zeros_(self.gate.bias)
        self.experts = nn.ModuleList([
            nn.Sequential(nn.Linear(hidden_dim, hidden_dim), nn.GELU(),
                          nn.Linear(hidden_dim, out_dim))
            for _ in range(n_experts)
        ])
    
    def forward(self, h, regime_feats):
        # h: [B, N, hidden_dim], regime_feats: [B, 4]
        weights = F.softmax(self.gate(regime_feats), dim=-1)  # [B, n_experts]
        expert_outs = torch.stack([e(h) for e in self.experts], dim=-1)  # [B, N, 3, n_experts]
        blended = (expert_outs * weights[:, None, None, :]).sum(-1)  # [B, N, 3]
        return blended, weights
```

### Step 3: Extract regime features from input

In the forward pass, extract global flow condition features:

```python
# re_norm = x[:, 0, 14] (normalized Re — check column index)
# aoa_norm = x[:, 0, 15] (or wherever AoA is stored)
# is_tandem = (x[:, :, 7].abs().max(dim=1).values > 0.01).float()  # nonzero dsdf2 = tandem
# gap_norm = x[:, :, 16].abs().max(dim=1).values  # gap magnitude proxy
# regime_feats = torch.stack([re_norm, aoa_norm, is_tandem, gap_norm], dim=-1)  # [B, 4]
```

Verify the actual column indices by inspecting x features in a debug run.

### Step 4: Add load-balancing loss

```python
if args.moe_output:
    # Maximize entropy of gate distribution to prevent collapse
    gate_entropy = -(weights * weights.log().clamp(min=-10)).sum(-1).mean()
    balance_loss = -args.moe_balance_weight * gate_entropy  # negative because we maximize entropy
    total_loss = total_loss + balance_loss
```

### Step 5: Run

```bash
CUDA_VISIBLE_DEVICES=0 python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/moe-output-s42" --wandb_group "moe-output-routing" \
  --moe_output --moe_n_experts 3 --moe_balance_weight 0.01 \
  [+ all baseline flags from reproduce command]

CUDA_VISIBLE_DEVICES=1 python train.py --agent alphonse --seed 73 \
  --wandb_name "alphonse/moe-output-s73" --wandb_group "moe-output-routing" \
  --moe_output --moe_n_experts 3 --moe_balance_weight 0.01 \
  [+ all baseline flags]
```

## Baseline

- **p_in:** 11.742 | **p_oodc:** 7.643 | **p_tan:** 27.874 | **p_re:** 6.419
- Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 --pressure_first --pressure_deep --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 --gap_stagger_spatial_bias --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 --te_coord_frame --wake_deficit_feature --re_stratified_sampling`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Flow-Regime MoE Output Routing — 2-Seed Run

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 13.4 | 13.1 | **13.250** | 11.742 | **+12.8%** ❌ |
| **p_oodc** | 9.0 | 8.8 | **8.900** | 7.643 | **+16.4%** ❌ |
| **p_tan** | 29.3 | 32.4 | **30.850** | 27.874 | **+10.7%** ❌ |
| **p_re** | 6.8 | 7.0 | **6.900** | 6.419 | **+7.5%** ❌ |

### Training commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent alphonse --seed 42 \
  --wandb_name "alphonse/moe-output-s42" --wandb_group "moe-output-routing" \
  --moe_output --moe_n_experts 3 --moe_balance_weight 0.01 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
# Seed 73: same, --seed 73, --wandb_name "alphonse/moe-output-s73"
```

### Run details

| | Seed 42 | Seed 73 |
|--|---------|---------|
| W&B run ID | skps8wwb | 60qrwzfe |
| Best epoch | 135 | 134 |
| Total epochs | 135 | 134 |
| Training time | 180.0 min | 180.3 min |
| Peak VRAM | 49.4 GB | 49.3 GB |
| MoE params | 112,920 | 112,920 |

### What happened

**Clear negative result — regressions across all metrics.** The MoE output correction head as designed hurts rather than helps performance.

**Analysis:**

1. **Fewer epochs**: Baseline reaches ~149 epochs; with the MoE (extra computation per forward), we reached 134-135 epochs (~10% fewer). This partially explains the regression, but not fully — the magnitude (+12.8% p_in, +16.4% p_oodc) exceeds what 10% fewer epochs alone would cause.

2. **Architecture interference**: The MoE is applied as an additive correction on top of the `pressure_first` output pathway. The pressure-first architecture already carefully conditions velocity on pressure. Adding a regime-conditioned correction that modifies all 3 channels simultaneously may disrupt this learned pressure→velocity dependency.

3. **Gate learning dynamics**: With zero-init experts, the gate starts at uniform routing and corrections start at zero. As the gate learns, it effectively adds a varying bias per regime. But the model already handles regime differences via:
   - `domain_velhead` (separate single/tandem output heads)
   - `pcgrad_3way` (gradient surgery between regimes)
   - `re_stratified_sampling` (upweighted extreme Re samples)
   Adding a 4th mechanism may cause conflicting signals rather than complementary ones.

4. **Seed 73 p_tan regression (+16.2%)**: Much worse than seed 42 (+5.0%), suggesting the MoE gate is routing tandem samples less reliably with a different initialization. High variance between seeds indicates the gate isn't finding stable routing.

5. **Gate collapse diagnostic**: The `moe/gate_entropy` metric in W&B can show whether experts are routing diversely. If entropy collapsed early, all experts are computing similar corrections — equivalent to a standard (non-MoE) correction head, but with more parameters.

**Implementation note**: The referenced paper arXiv:2508.21249 appears to be from 2025 (beyond my knowledge cutoff), so I could not verify its claims directly. The hypothesis that "regime-conditioned MoE improves on regime boundaries" may require different routing (hard dispatch, not soft blending) or different conditioning (per-node, not global) to match claimed gains.

### Suggested follow-ups

1. **Hard dispatch instead of soft blending**: Try top-1 gating where each sample routes entirely to one expert. Cleaner specialization, but risks load imbalance.
2. **Replace `domain_velhead` with MoE**: Currently both are active simultaneously, creating redundancy. Disable `domain_velhead` and let MoE handle the single/tandem split.
3. **Per-node conditioning**: Route on node-level features (curvature, distance to surface) rather than global regime features, since pressure patterns vary across the foil surface.
4. **Fewer interaction terms**: The model already handles regime diversity. The marginal benefit of an additional routing mechanism is likely small. Focus on other hypotheses.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2323 because ADVISOR: Closing — MoE output routing regresses all metrics (p_in +12.8%, p_oodc +16.4%, p_tan +10.7%, p_re +7.5%).

Your analysis is excellent. The key insight: the model already handles regime diversity through 3 mechanisms (domain_velhead, pcgrad_3way, re_stratified_sampling). Adding a 4th creates conflicting signals. The 10% epoch reduction from compute overhead makes it worse.

The seed-73 p_tan regression (+16.2% vs seed-42 +5.0%) confirms gate routing instability. Combined with prior evidence that MoE FFN routing starved experts (small dataset), **we can conclude that output-level MoE is not viable for our architecture/dataset size.**

Moving you to a fresh experiment..
